pkg/controller/node_availability.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"sync/atomic"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog/v2"
+	azClientSet "sigs.k8s.io/azuredisk-csi-driver/pkg/apis/client/clientset/versioned"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type ReconcileNodeAvailability struct {
+	client                client.Client
+	controllerSharedState *SharedState
+	azVolumeClient        azClientSet.Interface
+	kubeClient            kubernetes.Interface
+}
+
+var _ reconcile.Reconciler = &ReconcileNodeAvailability{}
+
+func (r *ReconcileNodeAvailability) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	n := &corev1.Node{}
+	err := r.client.Get(ctx, request.NamespacedName, n)
+
+	if errors.IsNotFound(err) {
+		return reconcile.Result{}, nil
+	}
+
+	if err == nil {
+		if !n.Spec.Unschedulable {
+			for _, taint := range n.Spec.Taints {
+				if taint.Effect == corev1.TaintEffectNoSchedule {
+					return reconcile.Result{Requeue: true}, nil
+				}
+			}
+
+			//Node is schedulable, proceed to attempt creation of replica attachment
+			if atomic.SwapInt32(&r.controllerSharedState.processingReplicaRequestQueue, 1) == 0 {
```

Comment:
Maybe what comes below fits better as a method under `VolumeReplicaRequestsPriorityQueue` receiver (e.g. `func (vq *VolumeReplicaRequestsPriorityQueue) tryCreateFailedReplicas(...)`).  That way, this procedure can be used elsewhere (e.g. after garbage collection)

Reply:
Good point, I've moved it now to the common.go